app/controllers/unattended_controller.rb

```diff
@@ -18,13 +21,24 @@ class UnattendedController < ApplicationController
   # all of our requests should be returned in text/plain
   after_action :set_content_type
 
-  # this actions is called by each operatingsystem post/finish script - it notify us that the OS installation is done.
   def built
-    logger.info "#{controller_name}: #{@host.name} is Built!"
+    logger.info "#{controller_name}: #{@host.name} is built!"
+    # clear possible previous errors
+    @host.build_errors = nil
     update_ip if Setting[:update_ip_from_built_request]
     head(@host.built ? :created : :conflict)
   end
 
+  def failed
+    return if preview? || !@host.build
+    logger.warn "#{controller_name}: #{@host.name} build failed!"
+    @host.build_errors = request.body.read(32768)
```

Comment:
What's the reason to limit reading to 2^15 bytes here? MySQL allows 2^16 bytes and SQLite as well where PostgreSQL is 1GB.

Reply:
DoS attack prevention, rack stack defaults to 32kB anyway I think. I don't want to make this configurable, we already have so many settings.

"32kB ought to be enough for everyone!" -- Bill Gates (it was not him actually, but nobody cares)

Reply:
I wonder if we should make it 64k to match mysql and sqlite but I don't know how much text we should expect.